### PR TITLE
WIP: redesigning error 404 page #210

### PR DESCRIPTION
**Pull request by mmlumba**
Screenshot from local:
screen shot 2016-12-09 at 9 54 06 am

Referencing issue #210 

PTAL @KHarshawat @shanear @lakohl

### CODE REVIEW

File: app/views/downloads/not_found.html.erb [hunk 1]
```diff
@@ -1,11 +1,6 @@
...
-    <p class="usa-alert-text">
-      We couldn't find that download.
-    </p>
-    <p>
-      <%= link_to "Try re-downloading that eFolder", root_path %>
-    </p>
-  </div>
+<div class="cf-app-msg-screen cf-app-segment cf-app-segment--alt">
+  <h1 class="cf-msg-screen-heading">Veteran ID Not Found</h1>
+  <h2 class="cf-msg-screen-deck">We weren't able to locate Veteran ID #123456789</h2>
```
Comment (shanear) [inline app/views/downloads/not_found.html.erb]: this number `#123456789` should not be hard coded.

File: spec/features/download_spec.rb [hunk 2]
```diff
@@ -186,7 +186,7 @@
     )
 
     visit download_path(another_user)
-    expect(page).to have_content("not found")
+    expect(page).to have_content("Not Found")
```
Comment (shanear) [inline spec/features/download_spec.rb]: once you add the veteran id for it not being hard coded, we should test that the proper vet id shows up on this page.

File: app/views/downloads/not_found.html.erb (hunk 1 above)
Comment (mmlumba) [inline app/views/downloads/not_found.html.erb]: @shanear What should be the source of the veteran ID?

**Comment by lakohl** [comment]
@mmlumba per my comment on #318 search again doesn't need to be underlined. Also our other CTAs are sentence case.

**Comment by mmlumba** [comment]
@KHarshawat @shanear @lakohl Styling update:
I modified the "not found" page so that when a user pulls up a _download_ ID that doesn't exist (or one they don't have access to), this is the page they see (in this example the invalid download ID is 10000):
screen shot 2016-12-16 at 11 18 30 am

Otherwise, if they type in a _veteran_ ID that doesn't exist at all, this is the page that they see:
screen shot 2016-12-16 at 11 18 52 am

But both should have the respective ID's on the pages, as well as fixing underlining and sentence casing.

**Comment by lakohl** [comment]
@mmlumba the Veteran ID Not Found page looks fine. I'm not sure what you mean by 'Download ID' and I don't think the user will either. Can you explain when this error would show? Also if the user does not have access to a file because they don't have the proper security permissions, we should show a message to that effect (not the Vet ID not found message). Thanks.

**Comment by mmlumba** [comment]
@lakohl So far the Download ID error shows if
1) There is a general 404 error (aka missing webpage)
2) The user accesses a download link that doesn't exist in the system

**Comment by lakohl** [comment]
Could you change the first two lines of copy for the 'Download ID Not Found' modal to
1. Uh-oh
2. Something went wrong...

The third line is fine as you have it!

**Comment by mmlumba** [comment]
@lakohl This is what the other page looks like now:
screen shot 2016-12-21 at 6 55 49 am
Could you please review?

@shanear Could you also do a technical review for this PR? The veteran ID is still not hard-coded on the veteran ID not found page, but for the other case in which there is a general 404 I changed the tests so it finds the text "search again" on that page.

**Comment by lakohl** [comment]
@mmlumba looking good. Could you take out the '!' on the first line and add '...' instead of '.' at the end of the second line. Thanks!

**Comment by mmlumba** [comment]
@lakohl PTAL below:
screen shot 2016-12-21 at 7 20 44 am

**Comment by lakohl** [comment]
Thanks! It's good to go.

File: app/models/download.rb [hunk 3]
```diff
@@ -7,7 +7,8 @@ class Download < ActiveRecord::Base
     packaging_contents: 4,
     complete_success: 5,
     complete_with_errors: 6,
-    vbms_connection_error: 7
+    vbms_connection_error: 7,
+    download_not_found: 8
```
Comment (amprokop) [inline app/models/download.rb]: naming nit: what is the name of the thing that is actually not found? is it the veteran's id #? should we name this `veteran_id_not_found` maybe?

File: app/views/downloads/_confirm.html.erb [hunk 4]
```diff
@@ -8,23 +8,32 @@
...
         <h3 class="usa-alert-heading">Couldn't find documents in eFolder</h3>
         <p class="usa-alert-text">
           eFolder Express could not find any documents in the eFolder with ID <%= @download.file_number %>. It's possible this eFolder does not exist.
         </p>
-        <p class="usa-alert-text"><%= link_to "Try again", root_path %></p>
+        <p class="usa-alert-text"><%= link_to "Search again", root_path %>.</p>
       </div>
     </div>
 
+  <% elsif @download.download_not_found? %>
```
Comment (amprokop) [inline app/views/downloads/_confirm.html.erb]: same as above— i feel there's a naming nit to be picked here. `veteran_id_not_found?` maybe?

File: app/assets/stylesheets/_main.scss [hunk 5]
```diff
@@ -68,6 +72,10 @@ table td {
   width: 1em;
 }
 
+.cf-msg-screen-text a {
+  text-decoration: none;
```
Comment (amprokop) [inline app/assets/stylesheets/_main.scss]: isn't this redundant if we have lines 31-33?

File: app/assets/stylesheets/_main.scss [hunk 6]
```diff
@@ -28,6 +28,10 @@
 }
 
 // Candidate for Caseflow Commons
+a {
+  text-decoration: none !important;
```
Comment (amprokop) [inline app/assets/stylesheets/_main.scss]: hmm. why do we need an an `important!` here? is it really the case that now and forever we will never want a link to have an underline / are we sure that we aren't accidentally removing an underline from a bunch of links that have it?

File: app/jobs/demo_get_download_manifest_job.rb [hunk 7]
```diff
@@ -55,12 +61,18 @@ def perform(download)
...
+    end
   end
 
   def check_and_raise_errors(demo)
-    fail VBMS::ClientError if demo[:error] && demo[:error_type] == "VBMS"
-    fail "no documents" if demo[:error] && demo[:error_type] == "NO_DOCUMENTS"
+    return unless demo[:error]
+    fail VBMS::ClientError if demo[:error_type] == "VBMS"
+    fail "no documents" if demo[:error_type] == "NO_DOCUMENTS"
+    fail "not found" if demo[:error_type] == "NOT_FOUND"
```
Comment (amprokop) [inline app/jobs/demo_get_download_manifest_job.rb]: similar to other naming nits — can we maybe make this error string more precisely describe the error case i.e. `ID_NOT_FOUND` or something?

File: app/views/downloads/_confirm.html.erb (hunk 4 above)
Comment (amprokop) [inline app/views/downloads/_confirm.html.erb]: or just `veteran_not_found`?

File: app/views/downloads/not_found.html.erb [hunk 8]
```diff
@@ -1,11 +1,6 @@
...
-    </p>
-    <p>
-      <%= link_to "Try re-downloading that eFolder", root_path %>
-    </p>
-  </div>
+<div class="cf-app-segment cf-app-segment--alt cf-app-msg-screen">
+  <h1 class="cf-msg-screen-heading">Uh oh</h1>
+  <h2 class="cf-msg-screen-deck">Something went wrong...</h2>
+  <p class="cf-msg-screen-text">Please check the Veteran ID # and <%= link_to "search again", root_path %>.</p>
+
```
Comment (amprokop) [inline app/views/downloads/not_found.html.erb]: we probably don't need this newline

File: app/views/downloads/page_not_found.erb [hunk 9]
```diff
@@ -0,0 +1,6 @@
+<div class="cf-app-msg-screen cf-app-segment cf-app-segment--alt">
+  <h1 class="cf-msg-screen-heading">Page not found</h1>
+  <h2 class="cf-msg-screen-deck">Oops! It looks like the page you're looking for doesn't exist.<br />
+    You might want to double check the URL for errors.</h2>
+
```
Comment (amprokop) [inline app/views/downloads/page_not_found.erb]: probably don't need this newline either

File: app/views/layouts/application.html.erb [hunk 10]
```diff
@@ -69,9 +69,6 @@
         <a class="cf-header-link" href="/">
           <h1 class="cf-logo"><span class="cf-logo-image"></span>eFolder Express</h1>
         </a>
-        <% if content_for?(:page_title) %>
-          <h2 id="page-title" class="cf-application-title"><%= yield :page_title %></h2>
```
Comment (amprokop) [inline app/views/layouts/application.html.erb]: hmm — why remove this?

**Comment by amprokop** [comment]
one more comment — it looks like this only works in demo right now, since we didn't modify `get_download_manifest_job.rb` yet, we only modified `demo_get_download_manifest_job.rb`. let me know if you need help working out how to use the real job service.

File: app/assets/stylesheets/_main.scss (hunk 6 above)
Comment (mmlumba) [inline app/assets/stylesheets/_main.scss]: @amprokop I've been told as per the style guide links within Caseflow should never have an underline. I'm working on caseflow commons so I can just make the style change there.

File: app/views/layouts/application.html.erb (hunk 10 above)
Comment (mmlumba) [inline app/views/layouts/application.html.erb]: @amprokop There was a styling issue that had the name of the user appear twice in the UI and it isn't supposed to look like that in the style guide, so I removed it.

File: app/assets/stylesheets/_main.scss (hunk 6 above)
Comment (amprokop) [inline app/assets/stylesheets/_main.scss]: ok, sounds good

File: app/views/layouts/application.html.erb (hunk 10 above)
Comment (amprokop) [inline app/views/layouts/application.html.erb]: 👍

**Comment by amprokop** [comment]
x-post from my Slack DM with @mmlumba, posting here so @shanear or @askldjd can correct me or jump in if necessary. @shanear might know the correct solution offhand.


> 
so if you don’t have a PIV card and a VA laptop, you can’t access the VA network
[11:59]  
to test out the real `get_download_manifest_job` or whatever, you need to be on the VA network

[11:59]  
(all the files in `/fakes` or with filenames prefixed with `demo` are just stubbed out versions of the real services)

[12:00]  
 ```class GetDownloadManifestJob < ActiveJob::Base
  queue_as :default

  def perform(download)
    vbms_documents = VBMSService.fetch_documents_for(download)

    if vbms_documents.empty?
      download.update_attributes!(status: :no_documents)
    else
      download_documents = DownloadDocuments.new(download: download, vbms_documents: vbms_documents)
      download_documents.create_documents
      download.update_attributes!(status: :pending_confirmation)
    end
  rescue VBMS::ClientError => e
    Rails.logger.error "#{e.message}\n#{e.backtrace.join("\n")}"
    Raven.capture_exception(e)
    download.update_attributes!(status: :vbms_connection_error)
  rescue
    download.update_attributes!(status: :no_documents)
  end

  def max_attempts
    1
  end
end
```
(edited)

Marian Lumba [12:00 PM]  
yeah that's what i know so far

[12:00]  
i'm just not sure how to do the testing through the VA network

Alex Prokop [12:00 PM]  
so you’ll have to add or change some condition here

[12:01]  
you’re trying to handle the case when the veteran isn’t found in the system, right?

Marian Lumba [12:02 PM]  
yes

Alex Prokop [12:04 PM]  
so i’m not sure what the response from `VBMSService` looks like when that’s the case

[12:04]  
it might be an error, or an empty array

[12:06]  
we do have a staging environment that hooks the local dev envirronment up to VA UAT, but you need to be on the VA network obv. i just got my VA laptop last week, so i’ve never used it and don’t quite know how. worth poking around in the README though.

[12:06]  
i know nemo has a PIV card now, he may have a VA laptop as well

[12:06]  
i think what you’d want to do to test this

[12:07]  
is get on the staging env and log the response from `VBMSService.fetch_documents_for(download)` when you’re searching for some veteran id that doesn’t exist on UAT

[12:08]  
if it’s an empty array, you’ll probably have to change the `vbms_documents.empty?` condition to `download.update_attributes!(status: :id_not_found)` or whatever

[12:09]  
if it’s an error, maybe you want to handle that specific error case, or change the generic `rescue` block to `download.update_attributes!(status: :id_not_found)`

[12:09]  
i think shane is out next week, right?

[12:09]  
he might know the answer to this offhand

[12:10]  
so my suggestion would be to ping nemo and see if he’s working today or tuesday or w/e

[12:10]  
and try to figure it out, but timebox it to ~2 hrs

[12:10]  
if you’re just spinning wheels, best to move on to something else and wait until shane or someone else is back

[12:10]  
maybe alan too

[12:10]  
what do you think of that plan?

Marian Lumba [12:12 PM]  
that works

[12:16]  
thanks alex!

Alex Prokop [12:39 PM]  
i’ll x-post this in the PR

[12:39]  
so shane or someone else can read or correct me if necessary

new messages
Marian Lumba [12:40 PM]  
ok 
>

**Comment by mmlumba** [comment]
@shanear PTAL at the latest changes.

File: app/views/downloads/page_not_found.erb [hunk 11]
```diff
@@ -0,0 +1,6 @@
+<div class="cf-app-msg-screen cf-app-segment cf-app-segment--alt">
```
Comment (shanear) [inline app/views/downloads/page_not_found.erb]: This file seems like it is unused. If that's the case can you delete it?

**Pull request merged**